### PR TITLE
buildah: add `heredoc` support for `RUN`, `COPY` and `ADD`

### PR DESCRIPTION
**Pull request by flouthoc**
Following PR is a attempt to add `Heredoc` support to buildah.

Once this PR is merged buildah is supposed to honor heredoc syntax while processing containerfiles

Expected syntax to work

```Dockerfile
FROM docker.io/library/python:latest
RUN <<EOF
echo "Hello" >> /hello
echo "World!" >> /hello
EOF

RUN python3 <<EOF
with open("/hello", "w") as f:
    print("Hello", file=f)
    print("Something", file=f)
EOF

RUN ls -a
RUN cat hello
```

#### What type of PR is this?

> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test 
> /kind feature
> /kind flake
> /kind other

#### Does this PR introduce a user-facing change?

```release-note
buildah: add heredoc support for RUN, COPY and ADD
```

Closes: https://github.com/containers/buildah/issues/3474

### CODE REVIEW

**Comment by flouthoc** [comment]
Needs: https://github.com/openshift/imagebuilder/pull/264

**Comment by flouthoc** [comment]
This needs: https://github.com/openshift/imagebuilder/pull/264

Other then the imagebuilder PR, following PR is ready to review.

@nalind @rhatdan @TomSweeneyRedHat @vrothberg @giuseppe @containers/buildah-maintainers PTAL

/hold wait till https://github.com/openshift/imagebuilder/pull/264

**Comment by TomSweeneyRedHat** [comment]
Changes here LGTM in general.  I assume you'll also be updating [the Containerfile man page in](https://github.com/containers/common/blob/main/docs/Containerfile.5.md) in c/common with the heredocs examples?

**Comment by flouthoc** [comment]
> Changes here LGTM in general. I assume you'll also be updating [the Containerfile man page in](https://github.com/containers/common/blob/main/docs/Containerfile.5.md) in c/common with the heredocs examples?

@TomSweeneyRedHat Yes I'll open a PR in `c/common` once this gets merged. As of now I am waiting for https://github.com/openshift/imagebuilder/pull/264

**Comment by flouthoc** [comment]
This PR is read for review.

File: imagebuildah/stage_executor.go
Comment (rhatdan) [inline imagebuildah/stage_executor.go:384]: This should be just after os.Create.

Comment (flouthoc) [reply imagebuildah/stage_executor.go]: fixed.

**Comment by Luap99** [comment]
I just took a quick looking at this after the demo, why are we importing all these protobuf files from builtkit? I understand that reuse is good but is there a way to do that without this useless protobuf code?

I am worried because a) that adds about 800k binary size and b) because it uses init() function it slows down every single command and allocates more memory, you check it by running buildah with `GODEBUG=inittrace=1`:
```
...
init github.com/gogo/protobuf/types @13 ms, 0.064 ms clock, 13264 bytes, 86 allocs
init github.com/containerd/typeurl @13 ms, 0.010 ms clock, 352 bytes, 4 allocs
init github.com/moby/buildkit/util/stack @13 ms, 0.13 ms clock, 87312 bytes, 58 allocs
init github.com/moby/buildkit/frontend/dockerfile/command @13 ms, 0.002 ms clock, 624 bytes, 2 allocs
init github.com/moby/buildkit/frontend/dockerfile/parser @13 ms, 0.079 ms clock, 23392 bytes, 262 allocs
...
```

To be clear I am not saying this is a blocker but we certainly should look into if we can prevent this from getting pulled in because IMO it should not slow down every single buildah/podman command as it is generally not needed. But I guess this is a lost battle within the go ecosystem.

**Comment by Luap99** [comment]
From a quick check
```diff
diff --git a/vendor/github.com/moby/buildkit/frontend/dockerfile/parser/errors.go b/vendor/github.com/moby/buildkit/frontend/dockerfile/parser/errors.go
index 9f28a5a2e..232ec366b 100644
--- a/vendor/github.com/moby/buildkit/frontend/dockerfile/parser/errors.go
+++ b/vendor/github.com/moby/buildkit/frontend/dockerfile/parser/errors.go
@@ -1,7 +1,6 @@
 package parser
 
 import (
-       "github.com/moby/buildkit/util/stack"
        "github.com/pkg/errors"
 )
 
@@ -41,10 +40,7 @@ func WithLocation(err error, location []Range) error {
        if errors.As(err, &el) {
                return err
        }
-       return stack.Enable(&ErrorLocation{
-               error:    err,
-               Location: location,
-       })
+       return err
 }
 
 func toRanges(start, end int) (r []Range) {
```

This is enough to prevent the huge import chain and protobuf deps.

**Comment by flouthoc** [comment]
@Luap99 I'm not sure if we can change this function like this, cause buildkit is using this so this patch will not be accepted at buildkit's end. Will it work if we move this to some other package different than `parser` and then buildkit uses it ?

**Review by nalind — CHANGES_REQUESTED** [review]
This needs conformance tests, since I'm concerned about whether or not we're adding (or not) newlines at end-of-file for COPY in a way that's consistent with expectations.

File: imagebuildah/stage_executor.go
Comment (flouthoc) [reply imagebuildah/stage_executor.go:651]: We are only using this here to detect if more than one heredoc is present or not, if there is any error in heredoc syntax imagebuilder fails much early with needed error so handling error here does will not have any effect since imagebuilder will fail early. I can add a test for bad case to verify this in CI 
```
Error: parsing main Dockerfile: /work/Dockerfile: EOF: unterminated heredoc
```

File: imagebuildah/stage_executor.go
Comment (flouthoc) [reply imagebuildah/stage_executor.go]: Writing it in `/var/tmp` on the host and then cleaning it up, i am wonder if there is any need to further sanitize the `file.Name` cause heredoc parser rejects funny strings for heredoc names like ones with `..` or `/` .

> But then if we're only creating them so that they can be copied into the rootfs, it might be preferable to generate an archive stream with archive/tar and extract that stream into the rootfs with copier.Put().

`s.builder.Add(` takes path as sources and does all the work related to perms and bit options if i'm gonna use `copier.Put` here, we will have to re-implement the logic or share it with `buildah.Add` therefore I thought using the same logic is more convenient, therefore creating a file and then cleaning it just after `copy` process.

File: imagebuildah/stage_executor.go
Comment (flouthoc) [reply imagebuildah/stage_executor.go]: Done, moved to `options.Mounts` which is cleaned after every step, I'll also add a test to verify this again.

**Comment by flouthoc** [comment]
> This needs conformance tests, since I'm concerned about whether or not we're adding (or not) newlines at end-of-file for COPY in a way that's consistent with expectations.

Conformance infra or `c/image` is unable to parse docker reference generated by docker daemon when a file with heredock is provided causing conformance test unable to process any heredoc files. I think `c/image`'s `docker daemon's` ParseReference needs to be fixed so it can read new buildkit images correctly. I am working on fixing `c/image` however I am manually verified output of `COPY` and `ADD`.

Sample containerfile which conformance test cannot read
```Dockerfile
# syntax=docker/dockerfile:1.4
FROM alpine
COPY <<robots.txt <<humans.txt /test/
(robots content)
robots.txt
(humans content)
humans.txt
```

File: imagebuildah/stage_executor.go
Comment (flouthoc) [reply imagebuildah/stage_executor.go]: Added a test for this.

**Comment by Luap99** [comment]
> @Luap99 I'm not sure if we can change this function like this, cause buildkit is using this so this patch will not be accepted at buildkit's end. Will it work if we move this to some other package different than `parser` and then buildkit uses it ? (That can still be accepted but will it reduce the protobuf dep ? )

Yeah that is what I am assuming too, the issue is that as soon `github.com/moby/buildkit/util/stack` is important you end up with the the protobuf packages as well which add ton of bloat.
I am not sure if buildkit upstream would care about something like that anyway, The easiest solution would be to have a special build tag for example or I guess we could fork it.
Either way I not saying we should block this but there i definitely room for improvements.

File: imagebuildah/stage_executor.go
Comment (nalind) [reply imagebuildah/stage_executor.go]: > Writing it in `/var/tmp` on the host and then cleaning it up, i am wonder if there is any need to further sanitize the `file.Name` cause heredoc parser rejects funny strings for heredoc names like ones with `..` or `/` .

path.Base(filepath.ToSlash(file.Name)), then.  We can't predict if that parser will become more forgiving later.

> > But then if we're only creating them so that they can be copied into the rootfs, it might be preferable to generate an archive stream with archive/tar and extract that stream into the rootfs with copier.Put().
> 
> `s.builder.Add(` takes path as sources and does all the work related to perms and bit options if i'm gonna use `copier.Put` here, we will have to re-implement the logic or share it with `buildah.Add` therefore I thought using the same logic is more convenient, therefore creating a file and then cleaning it just after `copy` process.

What are the expected permissions on a COPY heredoc?  0711 seems like an odd choice to hard wire, since that's execute for the owning group and everyone, but read-write-execute for the owner.

Comment (flouthoc) [reply imagebuildah/stage_executor.go]: > path.Base(filepath.ToSlash(file.Name)), then. We can't predict if that parser will become more forgiving later.
I added this change, but can add `copier.Put` if that is better.

File: imagebuildah/stage_executor.go
Comment (nalind) [inline imagebuildah/stage_executor.go:623]: This method doesn't need to be exported.

Comment (flouthoc) [reply imagebuildah/stage_executor.go]: fixed, I'll squash dev commits when PR is approved.

File: imagebuildah/stage_executor.go
Comment (flouthoc) [reply imagebuildah/stage_executor.go]: This is fixed in newer commit.

File: imagebuildah/stage_executor.go
Comment (flouthoc) [reply imagebuildah/stage_executor.go]: This is fixed in newer commit.

File: imagebuildah/stage_executor.go
Comment (flouthoc) [reply imagebuildah/stage_executor.go:632]: This is fixed in newer commit.

File: imagebuildah/stage_executor.go
Comment (flouthoc) [reply imagebuildah/stage_executor.go]: This is fixed in newer commit.

File: imagebuildah/stage_executor.go
Comment (flouthoc) [reply imagebuildah/stage_executor.go:637]: This is fixed in newer commit.

**Comment by flouthoc** [comment]
This PR needs a rebase after https://github.com/containers/buildah/pull/5165 since it uses a commit from the same PR.

File: imagebuildah/stage_executor.go
Comment (nalind) [inline imagebuildah/stage_executor.go:378]: Use `parse.GetTempDir()` here?

File: tests/conformance/testdata/Dockerfile.heredoc_copy
Comment (nalind) [inline tests/conformance/testdata/Dockerfile.heredoc_copy:4]: Unless there's a syntax reason it's not allowed, we should be testing COPY of a combination of heredocs and items from the context directory somewhere.

Comment (nalind) [reply tests/conformance/testdata/Dockerfile.heredoc_copy:4]: Maybe also with a `--from` to mix things up a little.

File: imagebuildah/stage_executor.go
Comment (nalind) [reply imagebuildah/stage_executor.go:378]: `os.Create()` is affected by the current process umask.  There needs to be a `tmpFile.Chmod()` in here somewhere.

File: tests/conformance/testdata/Dockerfile.heredoc_copy
Comment (flouthoc) [reply tests/conformance/testdata/Dockerfile.heredoc_copy]: Done.

**Comment by flouthoc** [comment]
@nalind Could you PTAL look again.

File: imagebuildah/stage_executor.go
Comment (nalind) [inline imagebuildah/stage_executor.go:379]: If the value doesn't have `<<` as a prefix, it can't have `<<-` as a prefix, so I don't think we need to be checking for both of these.

File: imagebuildah/stage_executor.go
Comment (nalind) [inline imagebuildah/stage_executor.go:410]: `tmpFile` is not closed in this branch.

File: imagebuildah/stage_executor.go
Comment (nalind) [inline imagebuildah/stage_executor.go:399]: After passing a pathname through `filepath.ToSlash()`, the result should be manipulated using the `path` package rather than `filepath`, since `filepath` expects the platform-specific path separator, which isn't `/` on Windows hosts.

File: tests/conformance/conformance_test.go
Comment (nalind) [inline tests/conformance/conformance_test.go:3022]: What values does it use?  Why is it okay to ignore the differences?

**Review by nalind — CHANGES_REQUESTED** [review]
Mostly nits and a conformance test question.

File: tests/conformance/conformance_test.go
Comment (flouthoc) [reply tests/conformance/conformance_test.go:3022]: Buildkit sets `uid` and `gid` to `1000` with `COPY` instruction in conformance test and `buildah/docker` keeps it 0, i suspect either problem is with buildkit client in conformance test or could be a buildkit issue, this happens on upstream as well so I was not able to spend time on it figure out root cause for it. Since this was not related to this PR hence I did not touch it for now  ( will investigate issue in a different PR )

**Issue reproducer for upstream:**

Following problem is with all the `COPY` ( where `uid` and `gid` are not modified ) related tests included in conformance_test.go

For example a try out example is
```diff
diff --git a/tests/conformance/conformance_test.go b/tests/conformance/conformance_test.go
index 6ec8f3837..e494cb116 100644
--- a/tests/conformance/conformance_test.go
+++ b/tests/conformance/conformance_test.go
@@ -2784,6 +2784,7 @@ var internalTestCases = []testCase{
        {
                name:       "copy-for-user", // flip side of issue #2518
                contextDir: "copy",
+               dockerUseBuildKit: true,
                dockerfileContents: strings.Join([]string{
                        `FROM alpine`,
                        `USER 66:66`,
```

and output fails

```console
[fl@fedora buildah]$ sudo GOROOT=/usr/local/go go test -v -timeout=30m -run TestConformance/copy-for-user ./tests/conformance
=== RUN   TestConformance
=== RUN   TestConformance/copy-for-user
    conformance_test.go:482: 
        	Error Trace:	/home/fl/heredoc-exp/buildah/tests/conformance/conformance_test.go:482
        	            				/home/fl/heredoc-exp/buildah/tests/conformance/conformance_test.go:333
        	            				/home/fl/heredoc-exp/buildah/tests/conformance/conformance_test.go:151
        	Error:      	Filesystem contents differ
        	Test:       	TestConformance/copy-for-user
        	Messages:   	File attributes in both versions have different values:
        	            	File:attr          Docker        buildah
        	            	/script:uid        1000          0
        	            	/script:gid        1000          0
    conformance_test.go:355: Context "/home/fl/heredoc-exp/buildah/tests/conformance/testdata/copy"
    conformance_test.go:367: Dockerfile contents:
        FROM alpine
        USER 66:66
        COPY /script /script
        (no final end-of-line)
=== RUN   TestConformance/copy-for-user-with-chown
--- FAIL: TestConformance (13.23s)
    --- FAIL: TestConformance/copy-for-user (6.97s)
    --- PASS: TestConformance/copy-for-user-with-chown (6.26s)
FAIL
FAIL	github.com/containers/buildah/tests/conformance	13.244s
FAIL
```

Comment (nalind) [reply tests/conformance/conformance_test.go:3022]: So BuildKit is retaining the uid/gid of the files in the context directory, which are owned by your unprivileged user, instead of setting them to 0?

Comment (nalind) [reply tests/conformance/conformance_test.go:3022]: I guess it's expecting the client to hand it an archive with ownerships set to 0:0 and go-dockerclient isn't doing that.

Comment (nalind) [reply tests/conformance/conformance_test.go:3022]: Opened https://github.com/containers/buildah/pull/5171 to generate the buildcontext archive we hand to dockerd with imagebuilder's helper, which sets ownerships in it to 0:0.

Comment (flouthoc) [reply tests/conformance/conformance_test.go]: Thanks I verified against #5171 and it works, I have removed `uid` and `gid` skips, will rebase after #5171

File: tests/conformance/conformance_test.go
Comment (nalind) [inline tests/conformance/conformance_test.go:3015]: This file comes from the build context, so I'd expect its datestamp to be consistent between different builds.

Comment (flouthoc) [reply tests/conformance/conformance_test.go]: correct, this skip is not needed.

File: imagebuildah/stage_executor.go
Comment (nalind) [inline imagebuildah/stage_executor.go:352]: What's the impact of this getting called in between adding heredocs and adding other content on the history and build caching?

Comment (flouthoc) [reply imagebuildah/stage_executor.go:352]: Refactored now digester does not restarts between heredoc and regular copy. Before this it was affecting history of the image keeping only the most recent copy in the history.

**Comment by flouthoc** [comment]
@nalind Could you PTAL again.

**Review by nalind — COMMENTED** [review]
Non-blocking nit, could probably use a rebase.

**Comment by TomSweeneyRedHat** [comment]
Grats on getting this in @flouthoc !  Nice work again!

**Pull request merged**
